File: windows/src/CodexQuotaTray.Core/Alerts/ResetCreditFingerprint.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using CodexQuotaTray.Core.Models;

namespace CodexQuotaTray.Core.Alerts;

public static class ResetCreditFingerprint
{
    public static string Create(ResetCreditView credit) => Create(
        credit.ResetType,
        credit.GrantedAtUtc,
        credit.ExpiresAtUtc,
        credit.Title);

    public static string Create(
        string? resetType,
        DateTimeOffset? grantedAtUtc,
        DateTimeOffset? expiresAtUtc,
        string? title)
    {
        var canonical = string.Join(
            '\u001f',
            Normalize(resetType),
            FormatTimestamp(grantedAtUtc),
            FormatTimestamp(expiresAtUtc),
            Normalize(title));
        return Convert.ToHexString(SHA256.HashData(Encoding.UTF8.GetBytes(canonical))).ToLowerInvariant();
    }

    private static string Normalize(string? value) => value?.Trim() ?? string.Empty;

    private static string FormatTimestamp(DateTimeOffset? value) =>
        value?.ToUnixTimeSeconds().ToString(CultureInfo.InvariantCulture) ?? string.Empty;
}

File: android/app/src/main/java/com/codexquotatray/android/alerts/ResetCreditFingerprint.kt
package com.codexquotatray.android.alerts

import com.codexquotatray.android.protocol.ResetCredit
import java.security.MessageDigest

object ResetCreditFingerprint {
    fun create(credit: ResetCredit): String = create(
        resetType = credit.resetType,
        grantedAt = credit.grantedAt,
        expiresAt = credit.expiresAt,
        title = credit.title,
    )

    fun create(
        resetType: String?,
        grantedAt: Long?,
        expiresAt: Long?,
        title: String?,
    ): String {
        val canonical = listOf(
            normalize(resetType),
            grantedAt?.toString().orEmpty(),
            expiresAt?.toString().orEmpty(),
            normalize(title),
        ).joinToString("\u001f")
        return MessageDigest.getInstance("SHA-256")
            .digest(canonical.toByteArray(Charsets.UTF_8))
            .joinToString("") { byte -> "%02x".format(byte) }
    }

    private fun normalize(value: String?): String = value?.trim().orEmpty()
}
